File: src/sync.js
/* eslint-disable func-names, no-console, no-await-in-loop, prefer-destructuring */

const fetch = require('node-fetch');
const path = require('path');
const flatten = require('lodash.flatten');
// const del = require('del');
const makeDir = require('make-dir');
const download = require('download');
const ProgressBar = require('progress');
const fs = require('fs');

const apiHost = 'https://api.bootcdn.cn';
const cdnHost = 'https://cdn.bootcss.com';

// JavaScript heap out of memory
let filePath;
let url;
let dirName;
const downloadOption = {};
let progressBar;

const sleep = function (delay) {
  return new Promise((resolve) => {
    setTimeout(resolve, delay);
  });
};

const getFilesByPackageName = async (name) => {
  const { assets = [] } = await fetch(`${apiHost}/libraries/${name}.min.json`).then(r => r.json());
  const pathss = assets
    // 暂时只取前十个版本, 作为 demo 展示
    .filter((_, ind) => (ind <= 5))
    .map(({ version, files = [] }) => files.map(fileName => ({
      name: `${name}_${version}_${fileName}`,
      filePath: path.resolve(__dirname, '../packages', name, version, fileName),
      url: `${cdnHost}/${name}/${version}/${fileName}`,
    })));
  return flatten(pathss);
};

const getFilesByUrl = async (thisLibList) => {
  for (let index = 0; index < thisLibList.length; index += 1) {
    filePath = thisLibList[index].filePath;
    url = thisLibList[index].url;
    dirName = path.dirname(filePath);
    downloadOption.filename = path.basename(filePath);
    if (!fs.existsSync(filePath)) {
      try {
        await makeDir(dirName);
        await sleep(10);
        await download(url, dirName, downloadOption);
      } catch (error) {
        console.log('Something error:', url);
        console.log(error);
      }
    }
    progressBar.tick();
  }
};

(async function () {
  const splitLine = '---------------------------------------------------';

  console.log(splitLine);
  console.log('Start process...');
  // console.log('Cleaning ./packages');
  // await del(path.resolve(__dirname, '../packages/**/*.*'));
  // console.log();

  console.log(splitLine);
  console.log('Start sync...');

  const libList = await fetch(`${apiHost}/names.min.json`).then(r => r.json());
  await sleep(50);

  let thisLibList;
  console.log('Total list items:', libList.length + 1);

  for (let i = 0; i < libList.length; i += 1) {
    thisLibList = await getFilesByPackageName(libList[i]);
    await sleep(20);
    progressBar = new ProgressBar(`${libList[i]} [:bar] :percent :elapsed s`, {
      complete: '=',
      incomplete: '-',
      width: 30,
      total: thisLibList.length,
    });
    await getFilesByUrl(thisLibList);
    await sleep(20);
  }
  console.log('Finished!!');
}());

File: src/sync.py
import threading
import requests
import os
import json
import time
from tqdm import tqdm
from multiprocessing import Process

CWD = os.getcwd() + '/py-packages/'

def download(link, filelocation):
  if os.path.exists(filelocation):
    return
  r = requests.get(link, stream=True)
  dirName = os.path.dirname(filelocation)
  if not os.path.exists(dirName):
    try:
      os.makedirs(dirName)
    except OSError as e:
      pass
  with open(filelocation, 'wb') as f:
    for chunk in r.iter_content(1024):
      if chunk:
        f.write(chunk)

def createNewDownloadThread(link, filelocation):
  # while threading.active_count() > 4:
  #   time.sleep(10)
  download_thread = Process(target=download, args=(link, filelocation))
  # download_thread = threading.Thread(target=download, args=(link,filelocation))
  download_thread.start()
  download_thread.join()

def getAssetsByLibName(ln):
  return requests.get('https://api.bootcdn.cn/libraries/' + ln + '.min.json').json()['assets']

def formatAssetsToDownload(libName, assets):
  fileList = []
  for aet in assets:
    libVersion = aet['version']
    for filePath in aet['files']:
      fileDict = {}
      fileDict['libName'] = libName
      fileDict['id'] = '%s_%s_%s' % (libName, libVersion, filePath)
      fileDict['fileOsPath'] = '%s%s/%s/%s' % (CWD, libName, libVersion, filePath)
      fileDict['url'] = 'https://cdn.bootcss.com/%s/%s/%s' % (libName, libVersion, filePath)
      fileList.append(fileDict)
  return fileList

def downloadFilesFromFileList(fileList):
  pbar = tqdm(fileList)
  for item in pbar:
    pbar.set_description("Asyncing %s" % item['libName'])
    createNewDownloadThread(item['url'], item['fileOsPath'])
    # download(item['url'], item['fileOsPath'])

libList = requests.get('https://api.bootcdn.cn/names.min.json').json()
for libName in libList:
  fileLists = formatAssetsToDownload(libName, getAssetsByLibName(libName))
  downloadFilesFromFileList(fileLists)
